gui/src/installer/step/mod.rs

```diff
@@ -111,6 +314,120 @@ fn bitcoind_default_address(network: &Network) -> String {
+    which::which("bitcoind").ok()
+}
+
+// Modified from https://github.com/RCasatta/bitcoind/blob/f047740d7d0af935ff7360cf77429c5f294cfd59/src/lib.rs#L435
+pub fn get_available_port() -> Result<u16, Error> {
+    // Perform multiple attempts to get a valid port.
+    for _ in 0..10 {
+        // Using 0 as port lets the system assign a port available.
+        let t = TcpListener::bind(("127.0.0.1", 0))
+            .map_err(|e| Error::CannotGetAvailablePort(e.to_string()))?;
+        let port = t
+            .local_addr()
+            .map(|s| s.port())
+            .map_err(|e| Error::CannotGetAvailablePort(e.to_string()))?;
+        if port_is_valid(&port) {
+            return Ok(port);
+        }
+    }
+    Err(Error::CannotGetAvailablePort(
+        "Exhausted attempts".to_string(),
+    ))
+}
+
+pub fn port_is_valid(port: &u16) -> bool {
+    !BITCOIND_DEFAULT_PORTS.contains(port)
+}
+
+impl Default for SelectBitcoindTypeStep {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl From<SelectBitcoindTypeStep> for Box<dyn Step> {
+    fn from(s: SelectBitcoindTypeStep) -> Box<dyn Step> {
+        Box::new(s)
+    }
+}
+
+impl SelectBitcoindTypeStep {
+    pub fn new() -> Self {
+        Self {
+            use_internal: false,
+        }
+    }
+}
+
+impl Step for SelectBitcoindTypeStep {
+    fn update(&mut self, message: Message) -> Command<Message> {
+        if let Message::SelectBitcoindType(msg) = message {
+            match msg {
+                message::SelectBitcoindTypeMsg::UseInternal(selected) => {
+                    self.use_internal = selected;
+                }
+            };
+            return Command::perform(async {}, |_| Message::Next);
+        };
+        Command::none()
+    }
+
+    fn apply(&mut self, ctx: &mut Context) -> bool {
+        if self.use_internal {
+            if ctx.internal_bitcoind_config.is_none() {
+                let mut conf = if let Ok(conf) =
+                    InternalBitcoindConfig::from_file(&internal_bitcoind_config_path(&ctx.data_dir))
+                {
+                    conf
+                } else {
+                    // TODO: Should the `else` only apply for "file not found" error?
+                    // In case of error while parsing existing file, we may overwrite values from other networks by creating new config.
+                    InternalBitcoindConfig::new()
+                };
+                // Insert entry for network if not present.
+                if conf.networks.get(&ctx.bitcoin_config.network).is_none() {
+                    let network_conf = match (get_available_port(), get_available_port()) {
+                        (Ok(rpc_port), Ok(p2p_port)) => {
+                            // In case ports are the same, user will need to click button again for another attempt.
+                            if rpc_port == p2p_port {
+                                return false;
+                            }
```

Comment:
Ah OK, I was thinking that after the function is called once, the port would still be available the next time the function is called, although even then it would probably be unlikely to return the same value. I've left it in for now just in case.